### PR TITLE
Implement hashbang grammar

### PR DESCRIPTION
**Pull request by ksh8281**

### CODE REVIEW

**Review by bbrto21 — APPROVED** [review]
LGTM

File: src/parser/esprima_cpp/esprima.cpp [hunk 1]
```diff
@@ -5260,6 +5260,28 @@ class Parser {
         ALLOC_TOKEN(token);
         while (true) {
             *token = this->lookahead;
+
+            if (UNLIKELY(token->type == Token::PunctuatorToken
```
Comment (bbrto21) [inline src/parser/esprima_cpp/esprima.cpp:5264]: What is this parsing algorithm based on?

File: src/parser/Lexer.h [hunk 2]
```diff
@@ -107,6 +107,8 @@ enum PunctuatorKind : uint8_t {
     LogicalNullishEqual,
     SubstitutionEnd,
 
+    HashBang,
```
Comment (bbrto21) [inline src/parser/Lexer.h:110]: Can you arrange these in alphabetical order? This is an opinion based on my personal preference, so you can ignore it :)

File: tools/test/test262/test262.py [hunk 3]
```diff
@@ -319,9 +320,16 @@ def GetDriverSource(self):
...
-
-    source = self.test
+        return self.source
+
+    source = self.source
+    # if test is hashbang test
+    if len(source) > 1 and (source[0] == '#' or source[0] == '\\'):
+      lines = source.splitlines()
+      if self.strict_mode:
+        lines.insert(1, 'var strict_mode = true;')
```
Comment (bbrto21) [inline tools/test/test262/test262.py:330]: Isn't your original intention to insert it into the zero index?

Comment (ksh8281) [reply tools/test/test262/test262.py:330]: Yes. 1 is intended number

File: src/parser/Lexer.h (hunk 2 above)
Comment (ksh8281) [reply src/parser/Lexer.h]: I reordered as alphabetical order

File: src/parser/esprima_cpp/esprima.cpp (hunk 1 above)
Comment (ksh8281) [reply src/parser/esprima_cpp/esprima.cpp:5274]: I added comment

File: src/parser/Lexer.h (hunk 2 above)
Comment (bbrto21) [reply src/parser/Lexer.h]: Thanks!

File: src/parser/esprima_cpp/esprima.cpp (hunk 1 above)
Comment (bbrto21) [reply src/parser/esprima_cpp/esprima.cpp:5274]: Thanks :+1:

**Review by bbrto21 — APPROVED** [review]
LGTM

File: tools/test/test262/test262.py [hunk 4]
```diff
@@ -319,9 +320,16 @@ def GetDriverSource(self):
...
+        lines.insert(1, '"use strict";')
```
Comment (clover2123) [inline tools/test/test262/test262.py:331]: Why did you add these two lines for strict mode?

Comment (ksh8281) [reply tools/test/test262/test262.py:331]: every test262 test cases doesn't have "use strict" directive at begining of its soruce code
so we have to append "use strict" directive if we want to test on strict-mode
you can see same thing on 335 line
```
    if self.strict_mode:
      source = '"use strict";\nvar strict_mode = true;\n' + source
```

Comment (clover2123) [reply tools/test/test262/test262.py:331]: I see :)

**Review by clover2123 — APPROVED** [review]
LGTM
